### ISSUE TITLE
Unable to resolve services in UnhandledExceptionDelegate, preventing logging or using it to do anything useful

### ISSUE THREAD

**Issue by kieranbenton** [graphql-dotnet/graphql-dotnet#1672]
## Description

Related to:
https://github.com/graphql-dotnet/graphql-dotnet/issues/1152

It is not possible to (at least easily) use UnhandledExceptionDelegate to resolve an ILogger<> (or any other service) and log out any exceptions that were thrown by execution. Without this capability UnhandledExceptionDelegate is pretty useless right now unfortunately.

If you can give me some pointers then I can try and fix this myself?

### Steps to reproduce

```
...
options.UnhandledExceptionDelegate = (unhandledExceptionContext) =>
{
    var log = unhandledExceptionContext.Services.GetRequiredService<ILogger<Startup>>();
    log.Error("Unhandled exception", unhandledExceptionContext.Exception);
};
...
```

```unhandledExceptionContext.Services``` (or equivalent) does not exist.

### Expected result

An instance of the ASP.NET service container should be exposed to this delegate so that services can be resolved to report errors.

### Actual result

Not possible.

### Environment

All, this is an API problem.

**Comment by sungam3r** [comment]
Whether or not it is easy to do depends on where you are trying. Consider this example:
```c#
publilc class MyMiddleware : GraphQLHttpMiddleware<MySchema>
{
  public MyMiddleware(IOptions<GraphQLOptions> options, ILogger<Startup> log, RequestDelegate next, PathString path, IGraphQLRequestDeserializer deserializer) : base(next, path, deserializer)
  {
    options.Value.UnhandledExceptionDelegate  = unhandledExceptionContext =>
    {
      log.Error("Unhandled exception", unhandledExceptionContext.Exception);
    };
  }
}
```
You can find `GraphQLHttpMiddleware` in the [server](https://github.com/graphql-dotnet/server/blob/develop/src/Transports.AspNetCore/GraphQLHttpMiddleware.cs#L27) project.
The idea is not to resolve dependencies yourself but allow DI-container to do its work. Configuring delegate can be done from the place where the dependency **is already resolved**.

**Comment by kieranbenton** [comment]
I can see that would work as a work around - thanks for that, and I will probably do this now as its achievable.

I should have been more clear though - my example is using the 'default' middleware and I'm trying to configure UnhandledExceptionDelegate within my Startup.ConfigureServices method (which since .NET Core 3 can no longer have an ILogger; let alone any other dependency; injected into it).

Wider example illustrating what I mean:
```
	services.AddGraphQL(options =>
	{
		options.EnableMetrics = true;
		options.ComplexityConfiguration = new ComplexityConfiguration()
		{
			MaxDepth = 25
		};
		options.ExposeExceptions = hostEnvironment.GetEnvironmentType().ToOnline() == EnvironmentType.Development;
		options.UnhandledExceptionDelegate = (unhandledExceptionContext) =>
		{
			var log = unhandledExceptionContext.Services.GetRequiredService<ILogger<Startup>>();
			log.Error("Unhandled exception", unhandledExceptionContext.Exception);
		})
	})
	.AddSystemTextJson(deserializerSettings => { }, serializerSettings => { }) 
	.AddUserContextBuilder<UserContextBuilder>() 
	.AddDataLoader() 
	.AddGraphTypes(typeof(MySchema));
```

So basically I'm suggesting that it should be possible to resolve services within the default middleware.

**Comment by sungam3r** [comment]
For what reason you can’t use inheritance? See one more [example](https://github.com/graphql-dotnet/server/blob/develop/samples/Samples.Server/GraphQLHttpMiddlewareWithLogs.cs).

> I'm suggesting that it should be possible to resolve services within the default middleware.

In fact, this is happening. `GraphQLHttpMiddleware` already [resolves](https://github.com/graphql-dotnet/server/blob/develop/src/Transports.AspNetCore/GraphQLHttpMiddleware.cs#L133) services from `HttpContext.RequestServices`.

> I'm trying to configure `UnhandledExceptionDelegate` within my `Startup.ConfigureServices` method.

This is probably not the best place to resolve dependencies. In any case, I would not expose `Services` to public API in any way. This feature has already been discussed many times, see #1151 and related issues.

**Comment by kieranbenton** [comment]
There's no reason why I can't use inheritance, apart from it seems like overkill just to allow me to log execution exceptions out. Like I said above right now that is likely what I will end up doing now I know this isn't supported.

My argument is, that if this API exists on the default middleware and service registration helpers, then it should be usable. Right now it isn't, as you can't get access to any services to do anything useful, barring say Console.WriteLine its pretty much a dead end.

If you don't want to pass an IServiceProvider through here to enable resolving some services - then it should be removed as its confusing, or at least have some documentation added to nudge users towards the solution you've proposed.

On a completely different note, whilst you are well within your rights to avoid resolution of services here (and yes, in a purist sense it is the service locator pattern), I think that ignores the practicality of how ASP.NET startup works and is a very different usecase to https://github.com/graphql-dotnet/graphql-dotnet/issues/1151 which I agree will quickly end up messy.

Great library btw!

**Comment by sungam3r** [comment]
>  then it should be removed as its confusing

**What** should be removed?

**Comment by sungam3r** [comment]
@kieranbenton I think that https://github.com/graphql-dotnet/server/pull/360 is what you really want. Note that this method of providing `IServiceProvider` is better than introducing any new `.Services` property, since it does not affect other layers of API.

**Comment by kieranbenton** [comment]
@sungam3r that PR is exactly what I was talking about - thank you.
